syne_tune/experiments/plotting.py

```diff
@@ -0,0 +1,664 @@
+    (both sides are inclusive), where ``None`` means the interval is open on
+    that side. This feature is useful to filter out results of erroneous
+    attempts.
+
+    If ``metadata_keys`` is given, it contains a list of keys into the
+    metadata. In this case, metadata values for these keys are extracted and
+    can be retrieved with :meth:`metadata_values`. In fact,
+    ``metadata_values(benchmark_name)`` returns a nested dictionary, where
+    ``result[key][setup_name]`` is a list of values. If
+    ``metadata_subplot_level`` is ``True`` and ``metadata_to_subplot`` is
+    given, the result structure is ``result[key][setup_name][subplot_no]``.
+    This should be set if different subplots share the same setup names,
+    since otherwise metadata values are only grouped by setup name.
+
+    :param experiment_names: Tuple of experiment names (prefixes, without the
+        timestamps)
+    :param setups: Possible values of setup names
+    :param num_runs: When grouping results w.r.t. benchmark name and setup
+        name, we should end up with this many experiments. See above
+    :param metadata_to_setup: See above
+    :param plot_params: Parameters controlling the plot. Can be overwritten
+        in :meth:`plot`. See :class:`PlotParameters`
+    :param metadata_to_subplot: See above. Optional
+    :param benchmark_key: Key for benchmark in metadata files. Defaults to
+        "benchmark". If this is ``None``, there is only a single benchmark,
+        and all results are merged together
+    :param with_subdirs: See above. Defaults to "*"
+    :param datetime_bounds: See above
+    :param metadata_keys: See above
+    :param metadata_subplot_level: See above. Defaults to ``False``
+    :param download_from_s3: Should result files be downloaded from S3? This
+        is supported only if ``with_subdirs``
+    :param s3_bucket: Only if ``download_from_s3 == True``. If not given, the
+        default bucket for the SageMaker session is used
+    """
+
+    def __init__(
+        self,
+        experiment_names: Tuple[str, ...],
+        setups: Iterable[str],
+        num_runs: int,
+        metadata_to_setup: MapMetadataToSetup,
+        plot_params: Optional[PlotParameters] = None,
+        metadata_to_subplot: Optional[MapMetadataToSubplot] = None,
+        benchmark_key: Optional[str] = "benchmark",
+        with_subdirs: Optional[Union[str, List[str]]] = "*",
+        datetime_bounds: Optional[DateTimeBounds] = None,
+        metadata_keys: Optional[List[str]] = None,
+        metadata_subplot_level: bool = False,
+        download_from_s3: bool = False,
+        s3_bucket: Optional[str] = None,
+    ):
+        if download_from_s3:
+            assert (
+                with_subdirs is not None
+            ), "Cannot download files from S3 if with_subdirs=None"
+            download_result_files_from_s3(experiment_names, s3_bucket)
+        result = create_index_for_result_files(
```

Comment:
How expensive is it to built this index? Would it make sense to dump this object, such that I can reload it when I run my plotting script multiple times, for example, because I adjust the parameters of plotting?

Reply:
Good point. In my experience so far, this is quite cheap. But we can also support storing it.